SmartDeviceLink/SDLBackgroundTaskManager.m

```diff
@@ -41,27 +41,38 @@ - (void)startBackgroundTask {
 
     __weak typeof(self) weakSelf = self;
     self.currentBackgroundTaskId = [[UIApplication sharedApplication] beginBackgroundTaskWithName:self.backgroundTaskName expirationHandler:^{
-        // We have ~1 second to do cleanup before ending the background task. If we take too long, the system will kill the app.
         __strong typeof(weakSelf) strongSelf = weakSelf;
-        SDLLogD(@"The background task %@ expired", strongSelf.backgroundTaskName);
-        if (strongSelf.taskEndedHandler != nil) {
-            SDLLogD(@"Waiting for cleanup to finish before ending the background task");
-            strongSelf.taskEndedHandler();
+        SDLLogD(@"The background task %@ is expiring.", strongSelf.backgroundTaskName);
+
+        // We have ~1 second to do cleanup before ending the background task. If we take too long, the system will kill the app.
+        if (strongSelf.taskExpiringHandler != nil) {
+            SDLLogD(@"Checking if subscriber wants to to perform some cleanup before ending the background task %@", strongSelf.backgroundTaskName);
+            BOOL waitForCleanupToFinish = strongSelf.taskExpiringHandler();
+            if (waitForCleanupToFinish) {
+                SDLLogD(@"Subscriber wants to clean up before ending the background task %@. Waiting...", self.backgroundTaskName);
+            } else {
+                SDLLogV(@"Subscriber does not want to perform cleanup. Ending the background task %@", strongSelf.backgroundTaskName);
+                [strongSelf endBackgroundTask];
+            }
         } else {
+            // No subscriber. Just end the background task.
+            SDLLogV(@"Ending background task %@", strongSelf.backgroundTaskName);
             [strongSelf endBackgroundTask];
         }
     }];
 
     SDLLogD(@"The %@ background task started with id: %lu", self.backgroundTaskName, (unsigned long)self.currentBackgroundTaskId);
 }
 
+- (void)expiredTaskCleanupFinished {
```

Comment:
What is the utility of this as opposed to `endBackgroundTask`. The only difference is that it checks that the `taskExpiringHandler` isn't `nil` first. I'm struggling to understand what the utility of that is. Can't the subscriber just call `endBackgroundTask`?

Reply:
Yes, that is the way I had it setup before but you seemed confused by how everything is interacting so I thought I would separate it out into its own function.

Reply:
I reverted to using `endBackgroundTask`.